src/computation/1_computation_parallel.cpp

```diff
@@ -327,3 +241,462 @@ void ComputationParallel::applyBoundaryValuesTop()
+        } else 
+        {
+            // u
+            std::array<double, discretization_->nCells()[1]+4> u_to_other_ghost_left;
+            for (int j = -2; j < discretization_->nCells()[1] +2; j++)
+            {
+                u_to_other_ghost_left[j+2] = u(0, j); // u_{0}
+            }
+            std::array<double, discretization_->nCells()[1]+4> other_ghost_to_u_left;
+
+            MPI_Recv(&other_ghost_to_u_left, 
+                     discretization_->nCells()[1]+4,
+                     MPI_DOUBLE,
+                     partitioning_.ownRankNo() - 1,
+                     )
+            MPI_Send(&u_to_other_ghost_left, 
+                     discretization_->nCells()[1]+4,
+                     MPI_DOUBLE,
+                     partitioning_.ownRankNo() - 1,
+                     )
+
+            for (int j = -2; j < discretization_->nCells()[1] +2; j++)
+            {
+                discretization_->u(-2, j) = other_ghost_to_u_left[j+2]; // u_{-2}
+            }
+
+            // v
+            std::array<double, discretization_->nCells()[1]+4> v_to_other_ghost_left;
+            for (int j = -2; j < discretization_->nCells()[1] +2; j++)
+            {
+                v_to_other_ghost_left[j+2] = v(0, j); // v_{0}
+            }
+            std::array<double, discretization_->nCells()[1]+4> other_ghost_to_v_left;
+
+            MPI_Recv(&other_ghost_to_v_left, 
+                     discretization_->nCells()[1]+4,
+                     MPI_DOUBLE,
+                     partitioning_.ownRankNo() - 1,
+                     )
+            MPI_Send(&v_to_other_ghost_left, 
+                     discretization_->nCells()[1]+4,
+                     MPI_DOUBLE,
+                     partitioning_.ownRankNo() - 1,
+                     )
+
+            for (int j = -2; j < discretization_->nCells()[1] +2; j++)
+            {
+                discretization_->v(-1, j) = other_ghost_to_v_left[j+2]; // v_{-1}
+            }
+        }
+    }
+}
+
+ComputationParallel::uvExchangeVertical()
+{
+    // the +4 in nCells+4 is from the consistent number of halo-cells in the staggered grid
+
+    // the even processes: send top, receive top, send bottom, receive bottom
+    if ((partitioning.ownRankNo() % 2) == 0)
+    {
+        // top
+        if (partitioning_.ownPartitionContainsTopBoundary())
+        {
+            applyBoundaryValuesTop();
+        } else 
+        {
+            // u
+            std::array<double, discretization_->nCells()[0]+4> u_to_other_ghost_top;
+            for (int i = -2; i < discretization_->nCells()[0] +2; i++)
+            {
+                u_to_other_ghost_left[i+2] = u(i, discretization_->nCells()[0] -1); // u_{n-1}
+            }
+            std::array<double, discretization_->nCells()[0]+4> other_ghost_to_u_top;
+
+            MPI_Send(&u_to_other_ghost_top, 
+                     discretization_->nCells()[0]+4,
+                     MPI_DOUBLE,
+                     partitioning_.ownRankNo() - 1,
```

Comment:
bei top und bottom muss doch ownRankNo je nachdem - oder + der gesamten ownRankNo in x-Richtung gerechnet werden